src/main/scala/subsystem/FrontBus.scala

```diff
@@ -20,5 +20,5 @@ class FrontBus(params: FrontBusParams)(implicit p: Parameters)
     with CanHaveBuiltInDevices
     with CanAttachTLMasters
     with HasTLXbarPhy {
-  attachBuiltInDevices(params)
+  val builtInDevices = attachBuiltInDevices(params)
```

Comment:
is there a reason this `val` can't go into the trait `CanHaveBuiltInDevices`? Maybe as a `def` or `lazy val`?

Reply:
this just seemed like the least intrusive way of adding it. the val can go in `CanHaveBuiltInDevices`, but then it would need to have access to a `params: HasBuiltInDeviceParams`. and I wasn't sure if `attachBuiltInDevices` was actually being called everywhere that `CanHaveBuiltInDevices` was being mixed in.

but it does look like the attach method is unconditionally called everywhere that the trait is mixed in. and the devices are already optionally attached based on the params. so yeah, it makes sense to me to have the `val` in `CanHaveBuiltInDevices` and just inline the `attachBuiltInDevices` method.

Reply:
but that seems like a cake-pattern and we're trying to avoid that?

Reply:
> but that seems like a cake-pattern and we're trying to avoid that?

We should all be clear on _why_ we are (or are not?) avoiding the cake pattern in future code, since it's rarely the case that we would conclude "X is always bad and we should never do X again". Not having a shared understanding of what the root problems actually are could lead to bandwagoning and scapegoating where we dump a solution without knowing why and end up building a new solution that runs into similar problems. My impression of why the way we have used the cake pattern has been problematic is because:

- We end up creating these [god classes](https://en.wikipedia.org/wiki/God_object) which have to know about every possible trait that we may want to use, causing every addition of a new trait or a change to a trait to have to propagate up to the god class. The reason we have the "CanHaveX" pattern is because we've turned the "might have or might not have feature X" into an optional thing everywhere.
- Parameterizing the concrete instances of the cake interfaces becomes challenging due to statement ordering issues and restrictions on constructors. Coupled with the fact that the Chisel DSL essentially requires people to write HDL code directly in the "constructor" of a Scala class, this makes it difficult to provide a consistent interface for concrete implementations of the cake interface.

See https://kubuszok.com/2018/cake-antipattern/ for a detailed blog post on the Scala cake pattern and why the author at least considers it to be an antipattern. See specifically "[Adding dependencies](https://kubuszok.com/2018/cake-antipattern/)", which provides a concrete example of a god class, and "[Initialization order](https://kubuszok.com/2018/cake-antipattern/)", which describes the ordering and initialization issues.

With these points in mind, my personal response would be that to best avoid the two problems above without also making a huge architectural change to the existing code would be to decouple the _interface_ from the _implementation_. We should not let the amount of typing required dictate what the interface should be. If we think that if a module says that it "can have built-in devices" should imply that we can retrieve said built-in devices under a uniform interface, then I think it is a good idea to define a `def`. However, if we think that different modules may want to expose them in different ways, then we should hold off from defining an interface.

Reply:
@richardxia I'm not sure what you are concretely suggesting, or what you need to know to make a concrete suggestion

Reply:
I need to know what a `CanHaveBuiltInDevices` conceptually represents, and I need to know how we intend for concrete implementations to use it as well as how consumers of the concrete implementations are meant to use it. Otherwise, I don't think it's possible to have a conversation about whether we should pull the `val builtInDevices = attachBuiltInDevices(params)` into the trait definition.

Let me flip this onto you: Is there a reason this val _should_ go into the trait CanHaveBuiltInDevices?

Reply:
I don't actually know what a `BuiltInDevice` is or the naming-conventions of these traits, but here's what I think:

I would think that `CanHaveBuiltInDevices` would have some field `builtInDevices: BuiltInDevices` so that consumers of the interface can know about and get some reference to the attached `Option[TLError/TLZero]`. right now it just defines a method `attachBuiltInDevices(params)` that can optionally be called to optionally attach devices based on `params`. If `CanHaveBuiltInDevices` doesn't define some interface for getting at the attached built-in devices, I don't quite understand the purpose of that trait. There's no reason `attachBuiltInDevices` needs to be implemented as a method inside a trait with self-type vs a regular function with a few extra arguments. and I don't think the `attachBuiltInDevices` method is helpful to consumers of the interface because if you call it from outside you risk instantiating and attaching devices multiple times, plus you'll also need a reference to the original `LazyScope` if you want the attach to actually happen in the same module that mixed in `CanHaveBuiltInDevices`.

I propose having an abstract `def builtInDevices: BuiltInDevices` in `CanHaveBuiltInDevices`, getting rid of the `TLBusWrapper` self-type, and adding a  `attachBuiltInDevices(params: HasBuiltInParams with HasTLBusParams, outwardNode: TLOutwardNode)` to the `BuiltInDevices` companion object. This way we can avoid adding another `lazy val` to the cake and still provide a useful interface for consumers.

Reply:
That sounds like a reasonable proposal to me, and I agree with the `trait` otherwise being useless since the `attach` method could just be in a standalone function and also that it is dangerous to call it twice.